File: client/clojure/src/cdisc_library_api/specs/sdtm_product.clj
(ns cdisc-library-api.specs.sdtm-product
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [cdisc-library-api.specs.sdtm-product-links :refer :all]
            [cdisc-library-api.specs.sdtm-class :refer :all]
            [cdisc-library-api.specs.sdtm-dataset :refer :all]
            )
  (:import (java.io File)))


(def sdtm-product-data
  {
   (ds/opt :name) string?
   (ds/opt :label) string?
   (ds/opt :description) string?
   (ds/opt :source) string?
   (ds/opt :effectiveDate) string?
   (ds/opt :registrationStatus) string?
   (ds/opt :version) string?
   (ds/opt :_links) sdtm-product-links-spec
   (ds/opt :classes) (s/coll-of sdtm-class-spec)
   (ds/opt :datasets) (s/coll-of sdtm-dataset-spec)
   })

(def sdtm-product-spec
  (ds/spec
    {:name ::sdtm-product
     :spec sdtm-product-data}))

File: client/scalaz/src/main/scala/org/openapitools/client/api/SdtmProduct.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime


import SdtmProduct._

case class SdtmProduct (
  name: Option[String],
label: Option[String],
description: Option[String],
source: Option[String],
effectiveDate: Option[String],
registrationStatus: Option[String],
version: Option[String],
links: Option[SdtmProductLinks],
classes: Option[List[SdtmClass]],
datasets: Option[List[SdtmDataset]])

object SdtmProduct {
  import DateTimeCodecs._

  implicit val SdtmProductCodecJson: CodecJson[SdtmProduct] = CodecJson.derive[SdtmProduct]
  implicit val SdtmProductDecoder: EntityDecoder[SdtmProduct] = jsonOf[SdtmProduct]
  implicit val SdtmProductEncoder: EntityEncoder[SdtmProduct] = jsonEncoderOf[SdtmProduct]
}
